File: substrate/chase_completion_collective_analysis_compose.py
"""Chase completion → collective analysis compose (pure).

After subagent chases complete, merge findings into draft/full analysis intent.
analysis_written, live_dispatched, pack_dispatched always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from substrate.collective_deep_research_merge import (
    CollectiveAnalysisIntent,
    CollectiveAnalysisMergeError,
    propose_collective_analysis_merge,
)

AnalysisMergeKind = Literal["draft_analysis", "full_analysis"]
VALID_STATUS = frozenset(("proposed", "open", "completed", "closed"))


class ChaseCompletionCollectiveAnalysisComposeError(ValueError):
    """Fail-closed validation for chase completion collective analysis."""


@dataclass(frozen=True)
class ChaseCompletionCollectiveAnalysisCompose:
    session_id: str
    parent_asset_id: str
    completed_slot_count: int
    selected_slot_ids: tuple[str, ...]
    analysis: CollectiveAnalysisIntent
    analysis_ready: bool
    analysis_written: bool
    live_dispatched: bool
    pack_dispatched: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "completed_slot_count": self.completed_slot_count,
            "selected_slot_ids": list(self.selected_slot_ids),
            "analysis": self.analysis.to_dict(),
            "analysis_ready": self.analysis_ready,
            "analysis_written": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "notes": list(self.notes),
            "authority": "chase_completion_collective_analysis_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ChaseCompletionCollectiveAnalysisComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_chase_completion_collective_analysis(
    *,
    session_id: object,
    parent_asset_id: object,
    slots: object,
    kind: object,
    operator_ack: object,
    extra_findings: object | None = None,
) -> ChaseCompletionCollectiveAnalysisCompose:
    """Compose pure collective analysis intent from completed chase slots."""
    if not isinstance(operator_ack, bool):
        raise ChaseCompletionCollectiveAnalysisComposeError(
            "operator_ack must be an explicit boolean"
        )
    if kind not in ("draft_analysis", "full_analysis"):
        raise ChaseCompletionCollectiveAnalysisComposeError(
            "kind must be draft_analysis or full_analysis"
        )

    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if not isinstance(slots, list) or len(slots) < 2:
        raise ChaseCompletionCollectiveAnalysisComposeError(
            "slots must be an array with at least 2 chase slots"
        )

    notes: list[str] = [
        "analysis_written=false — pure collective analysis intent only",
        "live_dispatched=false — no chase re-dispatch",
        "pack_dispatched=false — no pack execution",
    ]

    instances: list[dict[str, Any]] = []
    selected_slot_ids: list[str] = []
    completed_slot_count = 0
    seen: set[str] = set()

    for i, s in enumerate(slots):
        if not isinstance(s, dict):
            raise ChaseCompletionCollectiveAnalysisComposeError(
                f"slots[{i}] must be an object"
            )
        slot_id = _require_nonempty(s.get("slot_id"), field=f"slots[{i}].slot_id")
        if slot_id in seen:
            raise ChaseCompletionCollectiveAnalysisComposeError(
                f"duplicate slot_id: {slot_id}"
            )
        seen.add(slot_id)
        _require_nonempty(s.get("question_id"), field=f"slots[{i}].question_id")
        p = _require_nonempty(
            s.get("parent_asset_id"), field=f"slots[{i}].parent_asset_id"
        )
        if p != parent:
            raise ChaseCompletionCollectiveAnalysisComposeError(
                "all slots must share parent_asset_id"
            )
        status = s.get("status")
        if status not in VALID_STATUS:
            raise ChaseCompletionCollectiveAnalysisComposeError(
                f"slots[{i}].status invalid"
            )
        if status == "closed":
            notes.append(f"skipped closed slot {slot_id}")
            continue
        if status == "completed":
            completed_slot_count += 1

        findings: list[str] = []
        raw = s.get("findings")
        if raw is not None:
            if not isinstance(raw, list):
                raise ChaseCompletionCollectiveAnalysisComposeError(
                    f"slots[{i}].findings must be string[] when set"
                )
            for j, f in enumerate(raw):
                if not isinstance(f, str) or not f.strip():
                    raise ChaseCompletionCollectiveAnalysisComposeError(
                        f"slots[{i}].findings[{j}] must be non-empty string"
                    )
                findings.append(f.strip())

        selected_slot_ids.append(slot_id)
        inst: dict[str, Any] = {
            "instance_id": slot_id,
            "parent_asset_id": p,
            "status": status,
        }
        if s.get("body") is not None:
            inst["highlight"] = s.get("body")
        if findings:
            inst["findings"] = findings
        instances.append(inst)

    if len(instances) < 2:
        raise ChaseCompletionCollectiveAnalysisComposeError(
            "need ≥2 non-closed chase slots for collective analysis"
        )

    notes.append(
        f"selected_slots={len(selected_slot_ids)} · completed={completed_slot_count}"
    )

    try:
        analysis = propose_collective_analysis_merge(
            instances,
            kind=kind,
            operator_ack=operator_ack,
            extra_findings=extra_findings,
        )
    except CollectiveAnalysisMergeError as e:
        raise ChaseCompletionCollectiveAnalysisComposeError(str(e)) from e
    notes.extend(analysis.notes)

    analysis_ready = False
    if kind == "full_analysis":
        analysis_ready = (
            operator_ack is True
            and completed_slot_count == len(instances)
            and analysis.analysis_written is False
        )
        if not operator_ack:
            notes.append(
                "analysis_ready=false — full_analysis requires operator_ack"
            )
        elif completed_slot_count != len(instances):
            notes.append(
                "analysis_ready=false — full_analysis requires all selected slots completed"
            )
        else:
            notes.append(
                "analysis_ready=true — full analysis intent ready; analysis_written=false"
            )
    else:
        analysis_ready = (
            len(instances) >= 2 and analysis.analysis_written is False
        )
        notes.append(
            "analysis_ready=true — draft analysis intent ready; analysis_written=false"
            if analysis_ready
            else "analysis_ready=false"
        )

    if analysis.analysis_written is not False:
        raise ChaseCompletionCollectiveAnalysisComposeError(
            "invariant: analysis_written must remain false"
        )

    notes.extend(
        (
            "analysis_written=false",
            "live_dispatched=false",
            "pack_dispatched=false",
        )
    )

    return ChaseCompletionCollectiveAnalysisCompose(
        session_id=session,
        parent_asset_id=parent,
        completed_slot_count=completed_slot_count,
        selected_slot_ids=tuple(selected_slot_ids),
        analysis=analysis,
        analysis_ready=analysis_ready,
        analysis_written=False,
        live_dispatched=False,
        pack_dispatched=False,
        notes=tuple(notes),
        authority="chase_completion_collective_analysis_compose_advisory",
    )


def format_chase_completion_collective_analysis_summary(
    c: ChaseCompletionCollectiveAnalysisCompose,
) -> str:
    return (
        f"analysis_ready={c.analysis_ready} · kind={c.analysis.kind} · "
        f"slots={len(c.selected_slot_ids)} · completed={c.completed_slot_count} · "
        f"analysis_written=false · live_dispatched=false · pack_dispatched=false"
    )


__all__ = [
    "ChaseCompletionCollectiveAnalysisCompose",
    "ChaseCompletionCollectiveAnalysisComposeError",
    "compose_chase_completion_collective_analysis",
    "format_chase_completion_collective_analysis_summary",
]

File: apps/reading/src/api/chaseCompletionCollectiveAnalysisCompose.ts
/**
 * Chase completion → collective analysis compose (pure).
 *
 * Operator vision: after subagent deep-research chases complete, merge their
 * findings into a written analysis (draft or full) without live write.
 *
 * analysis_written always false.
 * live_dispatched always false.
 * pack_dispatched always false.
 */

import {
  proposeCollectiveAnalysisMerge,
  type AnalysisMergeKind,
  type CollectiveAnalysisIntent,
  type CompletedResearchInstance,
} from "./collectiveDeepResearchMerge";

export type ChaseSlotStatus = "proposed" | "open" | "completed" | "closed";

export interface CompletedChaseSlot {
  slot_id: string;
  question_id: string;
  parent_asset_id: string;
  status: ChaseSlotStatus;
  /** Operator/caller-supplied findings — never invented. */
  findings?: string[] | null;
  body?: string;
}

export interface ChaseCompletionCollectiveAnalysisInput {
  session_id: string;
  parent_asset_id: string;
  slots: CompletedChaseSlot[];
  kind: AnalysisMergeKind;
  operator_ack: boolean;
  extra_findings?: string[] | null;
}

export interface ChaseCompletionCollectiveAnalysisCompose {
  session_id: string;
  parent_asset_id: string;
  completed_slot_count: number;
  selected_slot_ids: string[];
  analysis: CollectiveAnalysisIntent;
  /**
   * True when ≥2 completed (for full) or ≥2 non-closed (for draft) slots
   * and analysis intent composed. Never writes analysis asset.
   */
  analysis_ready: boolean;
  analysis_written: false;
  live_dispatched: false;
  pack_dispatched: false;
  notes: string[];
  authority: "chase_completion_collective_analysis_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose pure collective analysis intent from completed chase slots.
 * Never writes analysis; never dispatches.
 */
export function composeChaseCompletionCollectiveAnalysis(
  input: ChaseCompletionCollectiveAnalysisInput,
): ChaseCompletionCollectiveAnalysisCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (input.kind !== "draft_analysis" && input.kind !== "full_analysis") {
    throw new Error("kind must be draft_analysis or full_analysis");
  }

  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  if (!Array.isArray(input.slots) || input.slots.length < 2) {
    throw new Error("slots must be an array with at least 2 chase slots");
  }

  const notes: string[] = [
    "analysis_written=false — pure collective analysis intent only",
    "live_dispatched=false — no chase re-dispatch",
    "pack_dispatched=false — no pack execution",
  ];

  const instances: CompletedResearchInstance[] = [];
  const selected_slot_ids: string[] = [];
  let completed_slot_count = 0;
  const seenSlot = new Set<string>();

  for (let i = 0; i < input.slots.length; i++) {
    const s = input.slots[i];
    if (!s || typeof s !== "object") {
      throw new Error(`slots[${i}] must be an object`);
    }
    const slot_id = requireNonEmpty(s.slot_id, `slots[${i}].slot_id`);
    if (seenSlot.has(slot_id)) {
      throw new Error(`duplicate slot_id: ${slot_id}`);
    }
    seenSlot.add(slot_id);
    requireNonEmpty(s.question_id, `slots[${i}].question_id`);
    const p = requireNonEmpty(
      s.parent_asset_id,
      `slots[${i}].parent_asset_id`,
    );
    if (p !== parent_asset_id) {
      throw new Error("all slots must share parent_asset_id");
    }
    if (
      s.status !== "proposed" &&
      s.status !== "open" &&
      s.status !== "completed" &&
      s.status !== "closed"
    ) {
      throw new Error(`slots[${i}].status invalid`);
    }
    if (s.status === "closed") {
      notes.push(`skipped closed slot ${slot_id}`);
      continue;
    }
    if (s.status === "completed") {
      completed_slot_count += 1;
    }

    const findings: string[] | undefined = [];
    if (s.findings != null) {
      if (!Array.isArray(s.findings)) {
        throw new Error(`slots[${i}].findings must be string[] when set`);
      }
      for (let j = 0; j < s.findings.length; j++) {
        const f = s.findings[j];
        if (typeof f !== "string" || !f.trim()) {
          throw new Error(
            `slots[${i}].findings[${j}] must be non-empty string`,
          );
        }
        findings.push(f.trim());
      }
    }

    selected_slot_ids.push(slot_id);
    instances.push({
      instance_id: slot_id,
      parent_asset_id: p,
      status: s.status,
      highlight: s.body,
      findings: findings.length > 0 ? findings : undefined,
    });
  }

  if (instances.length < 2) {
    throw new Error(
      "need ≥2 non-closed chase slots for collective analysis",
    );
  }

  notes.push(
    `selected_slots=${selected_slot_ids.length} · completed=${completed_slot_count}`,
  );

  let analysis: CollectiveAnalysisIntent;
  try {
    analysis = proposeCollectiveAnalysisMerge(instances, {
      kind: input.kind,
      operator_ack: input.operator_ack,
      extra_findings: input.extra_findings,
    });
  } catch (e) {
    throw e instanceof Error ? e : new Error(String(e));
  }
  notes.push(...analysis.notes);

  // analysis_ready: compose succeeded AND for full, all selected completed
  let analysis_ready = false;
  if (input.kind === "full_analysis") {
    analysis_ready =
      input.operator_ack &&
      completed_slot_count === instances.length &&
      analysis.analysis_written === false;
    if (!input.operator_ack) {
      notes.push("analysis_ready=false — full_analysis requires operator_ack");
    } else if (completed_slot_count !== instances.length) {
      notes.push(
        "analysis_ready=false — full_analysis requires all selected slots completed",
      );
    } else {
      notes.push(
        "analysis_ready=true — full analysis intent ready; analysis_written=false",
      );
    }
  } else {
    // draft: ack optional for scaffold but we mirror merge readiness
    analysis_ready = instances.length >= 2 && analysis.analysis_written === false;
    notes.push(
      analysis_ready
        ? "analysis_ready=true — draft analysis intent ready; analysis_written=false"
        : "analysis_ready=false",
    );
  }

  if (analysis.analysis_written !== false) {
    throw new Error("invariant: analysis_written must remain false");
  }

  notes.push("analysis_written=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");

  return {
    session_id,
    parent_asset_id,
    completed_slot_count,
    selected_slot_ids,
    analysis,
    analysis_ready,
    analysis_written: false,
    live_dispatched: false,
    pack_dispatched: false,
    notes,
    authority: "chase_completion_collective_analysis_compose_advisory",
  };
}

export function formatChaseCompletionCollectiveAnalysisSummary(
  c: ChaseCompletionCollectiveAnalysisCompose,
): string {
  return (
    `analysis_ready=${c.analysis_ready} · kind=${c.analysis.kind} · ` +
    `slots=${c.selected_slot_ids.length} · completed=${c.completed_slot_count} · ` +
    `analysis_written=false · live_dispatched=false · pack_dispatched=false`
  );
}
